docs: Document AWS system credentials (federated authentication) (#5081)
n8n-io/n8n-docs: docs/integrations/builtin/credentials/aws.md

* **Principle of least privilege:** Grant only the permissions needed for specific tasks
* **Audit trail:** Better tracking of who accessed what resources

> n8n uses the official AWS SDK to make the STS `AssumeRole` call, and requests fresh temporary credentials when it signs node requests, so you don't need to manage their expiry. This differs from the **Temporary security credential** option on the AWS (IAM) credential, where you enter a static session token that n8n can't renew: when that token expires, you must enter a new one.

Reviewer: ```suggestion
n8n uses the official AWS SDK to make the STS `AssumeRole` call. It requests fresh temporary credentials when it signs node requests, so you don't need to manage their expiry. This differs from the **Temporary security credential** option on the AWS (IAM) credential, where you enter a static session token that n8n can't renew: when that token expires, you must enter a new one.
```

Reviewer 2: Done, applied as suggested.

##### Option 1: Use system credentials (recommended for server deployments) <a href="#option-1-use-system-credentials-recommended-for-server-deployments" id="option-1-use-system-credentials-recommended-for-server-deployments"></a>

Turn on **Use System Credentials** if the environment your n8n server runs in already has an AWS identity. n8n then discovers credentials for the `AssumeRole` call automatically, without storing static keys in n8n. n8n tries these sources in order and uses the first one that returns credentials:

> 1. Environment variables (`AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and optionally `AWS_SESSION_TOKEN`)

Reviewer: Between this list and the "Sources" in the table under "Supported credential sources", it would be preferable to use identical naming if possible, unless there is a legitimate difference in naming between the two contexts?

Reviewer 2: Good catch, I made the two lists match. Could you check I picked the right direction: I changed the table's Source column to match this list, rather than the other way round, since the list has no separate platform column to lean on.

2. EKS IAM Roles for Service Accounts (IRSA)
3. EKS Pod Identity
4. ECS or Fargate task role

> 5. EC2 instance profile

Reviewer: ```suggestion
5. EC2 Instance profile
```

Reviewer 2: Reverted this to lowercase in 32db8bf78, since cubic flagged the capital as inconsistent with the sources table, the trust policy prose, and item 4 in this same list.

##### Option 1: Use system credentials (recommended for server deployments) <a href="#option-1-use-system-credentials-recommended-for-server-deployments" id="option-1-use-system-credentials-recommended-for-server-deployments"></a>

Turn on **Use System Credentials** if the environment your n8n server runs in already has an AWS identity. n8n then discovers credentials for the `AssumeRole` call automatically, without storing static keys in n8n. n8n tries these sources in order and uses the first one that returns credentials:

1. Environment variables (`AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and optionally `AWS_SESSION_TOKEN`)
2. EKS IAM Roles for Service Accounts (IRSA)
3. EKS Pod Identity
4. ECS or Fargate task role

> 5. EC2 Instance profile

Reviewer 2: Applied in 32db8bf78.

* Use unique External IDs for each credential to prevent unauthorized access.
* Rotate the STS credentials used for role assumption.
* Apply the principle of least privilege to both the assuming credentials and the target role.
* On shared instances, keep system credentials access disabled. Refer to [Using AWS system credentials](#using-aws-system-credentials-federated-authentication).

## Using AWS system credentials (federated authentication)

This section is for administrators of self-hosted n8n instances.

{% hint style="info" %}
System credentials access is for self-hosted n8n only. It's not available on n8n Cloud.
{% endhint %}

When n8n runs on AWS infrastructure, the infrastructure itself can provide an AWS identity. There are no access keys to create, store, or rotate. This is the standard way to authenticate workloads on EKS, ECS, and EC2, and security teams often require it ("zero static keys").

> n8n reads these **system credentials** with the official AWS SDK credential providers, and uses them as the starting credentials for the **AWS (Assume Role)** credential when **Use System Credentials** is turned on. This works for all nodes that accept the AWS (Assume Role) credential, including the AWS Bedrock AI nodes.

Reviewer 2: Verified this section against the implementation and it's accurate throughout. Two gaps worth closing before merge, both about the same thing: a Role ARN is always mandatory.

There's no path where the server's identity authenticates a node directly. `assertValidRoleCredentials` in `packages/nodes-base/credentials/common/aws/utils.ts` throws `Role ARN is required when assuming a role.` unconditionally, and `getSystemCredentials()` is only ever reached as the master credential for an STS `AssumeRole` call.

So an EKS user whose IRSA role already carries the permissions they need must still supply a Role ARN, pointing either at a second role or at the IRSA role itself. Self-assume requires adding the role's own ARN to its trust policy, which is the step people most often miss.

```suggestion
n8n reads these **system credentials** with the official AWS SDK credential providers, and uses them as the starting credentials for the **AWS (Assume Role)** credential when **Use System Credentials** is turned on. This works for all nodes that accept the AWS (Assume Role) credential, including the AWS Bedrock AI nodes.

{% hint style="info" %}
System credentials are always the starting point for a role assumption, never the identity n8n calls AWS with directly. You must still enter a **Role ARN**, and that role must trust the identity n8n runs as.

If the role n8n already runs as has the permissions your workflows need, enter that same role's ARN and allow it to assume itself. Refer to [Trust policy for system credentials](#trust-policy-for-system-credentials).
{% endhint %}
```

Reviewer 2: Applied in 2f6f69b21.

| 5 | Instance profile | EC2 instance metadata service (IMDSv2) | Amazon EC2 |

Things to keep in mind:

* AWS injects the detection variables for sources 2 to 4 automatically when you configure the platform feature. You don't set them yourself.
* If you set `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` on the container, they take precedence over IRSA, Pod Identity, and instance roles. Remove them if you want the federated source to be used.
* n8n resolves system credentials when it needs them, so rotated or short-lived source credentials (for example, the IRSA token that Kubernetes rotates) are picked up automatically.
* Reading AWS CLI profiles from `~/.aws` (`AWS_PROFILE`) is intentionally not supported.

#### Amazon EKS with IRSA

[IAM Roles for Service Accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) associates an IAM role with the Kubernetes service account that runs n8n. Annotate the service account with the role ARN (`eks.amazonaws.com/role-arn`). EKS then injects `AWS_ROLE_ARN` and `AWS_WEB_IDENTITY_TOKEN_FILE` into the pod, and n8n exchanges the token with the regional STS endpoint.

#### Amazon EKS with Pod Identity

> [EKS Pod Identity](https://docs.aws.amazon.com/eks/latest/userguide/pod-identities.html) is the newer alternative to IRSA. Install the Pod Identity Agent add-on and create a pod identity association between the n8n service account and an IAM role. No OIDC provider setup needed.

Reviewer: ```suggestion
[EKS Pod Identity](https://docs.aws.amazon.com/eks/latest/userguide/pod-identities.html) is the newer alternative to IRSA. Install the Pod Identity Agent add-on and create a pod identity association between the n8n service account and an IAM role. No OIDC provider setup is needed.
```

Reviewer 2: Done, applied as suggested.

#### Amazon ECS and Fargate

Assign a [task role](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html) to the n8n task definition.

#### Amazon EC2

> Attach an [instance profile](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2_instance-profiles.html) to the instance running n8n. n8n uses IMDSv2, so instances with IMDSv1 disabled work.

Reviewer 2: Follow-on from the Role ARN comment: the page has no trust policy example for the federated path, and the existing one under **Setup Steps** uses `"AWS": "arn:aws:iam::SOURCE-ACCOUNT:root"`. That's right for manual STS keys but wrong here, where the `Principal` must be the IRSA, Pod Identity, task, or instance role.

An `AccessDenied` on `sts:AssumeRole` is the most likely failure right after an admin enables the flag, so it's worth a short subsection. Suggesting it here, after the four platform subsections.

````suggestion
Attach an [instance profile](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2_instance-profiles.html) to the instance running n8n. n8n uses IMDSv2, so instances with IMDSv1 disabled work.

### Trust policy for system credentials

The role you enter as **Role ARN** must trust the identity n8n runs as: the IRSA role, the Pod Identity role, the ECS task role, or the EC2 instance role. Use that role's ARN as the `Principal`, not the account root.

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::123456789012:role/n8n-service-account-role"
      },
      "Action": "sts:AssumeRole",
      "Condition": {
        "StringEquals": {
          "sts:ExternalId": "your-unique-external-id"
        }
      }
    }
  ]
}
```

To find the ARN to use as the `Principal`, run `aws sts get-caller-identity` from inside the n8n container or instance.

If the role n8n runs as already has the permissions your workflows need, use that same role as the **Role ARN** and add its own ARN to its trust policy as shown above. AWS requires this even when a role assumes itself.

The role n8n runs as also needs `sts:AssumeRole` permission on the target role. Grant it on the specific roles workflows should use rather than on `*`.
````

Reviewer 2: Applied in 2f6f69b21 as a new `Trust policy for system credentials` section, plus a pointer on the existing Setup Steps policy noting it covers Option 2.

|---|---|---|
| `N8N_AWS_SYSTEM_CREDENTIALS_ACCESS_ENABLED` | `false` | Allow the **Use System Credentials** option on AWS (Assume Role) credentials to read the server's AWS identity. |
| `N8N_AWS_SYSTEM_CREDENTIALS_SDK_SOURCES` | `all` | Transitional, available from n8n 2.29.0. Controls which system-credential sources resolve through the AWS SDK rather than n8n's legacy resolver. Accepts `all`, `none`, or a comma-separated subset of `environment`, `roleForServiceAccount`, `podIdentity`, `containerMetadata`, `instanceMetadata`. Only change this if AWS authentication broke after an upgrade. n8n removes this flag in a future release. |
| `N8N_AWS_LEGACY_SIGNER` | `false` | Transitional, available from n8n 2.30.0. Set to `true` to sign AWS requests with the legacy `aws4` signer instead of AWS's Signature V4 implementation. Only use as a rollback if AWS requests fail after an upgrade. n8n removes this flag in a future release. |

### Troubleshooting

* **"Access to AWS system credentials disabled, contact your administrator."**: The **Use System Credentials** option is on, but `N8N_AWS_SYSTEM_CREDENTIALS_ACCESS_ENABLED` isn't set to `true` on the server.

> * **The wrong identity is used**: Check the source order above. Environment variables beat IRSA, which beats Pod Identity, ECS, and EC2. A stray `AWS_ACCESS_KEY_ID` in the container is the most common cause.

Reviewer: ```suggestion
* **The wrong identity is used**: Environment variables beat IRSA, which beats Pod Identity, ECS, and EC2. A stray `AWS_ACCESS_KEY_ID` in the container is the most common cause. See [Supported credential sources](#supported-credential-sources)."
```

Reviewer 2: Applied, thanks. Two small extra changes on top of your suggestion, could you take a look: I dropped a stray trailing quote mark that was in the suggested text, and I used "Refer to" instead of "See" to match the wording used elsewhere on this page.

Reviewer: Great, thanks

> * **Credential test fails on EKS**: Verify the service account annotation (IRSA) or the pod identity association, and confirm the pod received the injected environment variables (`kubectl exec <pod-name> -- env | grep AWS`).

Reviewer 2: One more troubleshooting entry to pair with the trust policy suggestion above, covering the failure an admin is most likely to hit first.

```suggestion
* **Credential test fails on EKS**: Verify the service account annotation (IRSA) or the pod identity association, and confirm the pod received the injected environment variables (`kubectl exec <pod-name> -- env | grep AWS`).
* **`AccessDenied` when calling `sts:AssumeRole`**: n8n found system credentials, but the target role rejected them. Check that the role's trust policy names the identity n8n runs as, and that the **External ID** matches the policy's condition. Refer to [Trust policy for system credentials](#trust-policy-for-system-credentials).
```

Reviewer 2: Applied in 2f6f69b21.
